src/runtime_src/core/tools/xbutil2/OO_P2P.cpp

```diff
@@ -17,12 +17,244 @@
 // Local - Include Files
 #include "OO_P2P.h"
 #include "tools/common/XBUtilities.h"
+#include "core/common/error.h"
+#include "core/common/unistd.h"
+#include "core/common/memalign.h"
+#include "core/common/device.h"
+#include "core/common/query_requests.h"
+#include "experimental/xrt-next.h"
+#include "experimental/xrt_bo.h"
+#include "xrt.h"
+#include "xclbin.h"
 namespace XBU = XBUtilities;
 
 // 3rd Party Library - Include Files
+#include <boost/program_options.hpp>
+#include <boost/algorithm/string/join.hpp>
+namespace po = boost::program_options;
 
 // System - Include Files
 #include <iostream>
+#include <cctype>
+
+#ifdef _WIN32
+# pragma warning(disable : 4245)
+#endif
+
+namespace {
+
+enum class action_type { enable, disable, validate };
+
+action_type
+string2action(std::string str)
+{
+  std::transform(str.begin(), str.end(), str.begin(),
+                 [](auto c) { return static_cast<char>(std::tolower(c)); });
+
+  if (str == "enable")
+    return action_type::enable;
+  else if (str == "disable")
+    return action_type::disable;
+  else if (str == "validate")
+    return action_type::validate;
+  else
+    throw xrt_core::generic_error(EINVAL, "Invalid p2p action '" + str + "'");
+}
+
+static auto
+p2p_config(xrt_core::device* device)
+{
+  try {
+    auto cfg = xrt_core::device_query<xrt_core::query::p2p_config>(device);
+    return cfg;
+  }
+  catch (const std::exception&) {
+    throw xrt_core::system_error(ENOTSUP, "p2p is not supported");
+  }
+}
+
+static void
+p2p_enabled_or_error(xrt_core::device* device)
+{
+  auto cfg = p2p_config(device);
+
+  long long bar = -1;
+  long long rbar = -1;
+  long long remap = -1;
+  long long exp_bar = -1;
+
+  for (auto& str : cfg) {
```

Comment:
Looking at the code it is evident that is is iterating over a collection (most likely a vector) of strings, where the string format is <key>:<value>.  

Could you add a comment with example "sample" string that is being parsed to assist the next developer in understanding what is being done.

Reply:
I have this function as check_p2p_config in XBUtilities

Reply:
Yes, I didn't notice this function in XBUtilities.  However, I want exceptions, so we will have to defer consolidating this.